azurelinuxagent/common/utils/textutil.py

```diff
@@ -396,3 +398,20 @@ def format_memory_value(unit, value):
         raise TypeError('Value must be convertible to a float')
 
     return int(value * units[unit])
+
+def str_to_encoded_ustr(s, encoding='utf-8'):
+    from azurelinuxagent.common.version import PY_VERSION_MAJOR
```

Comment:
can we move the import to the top of the file (easier to see module dependencies when reading the code)

Reply:
I tried doing that initially but it started giving me a lot of cyclic dependency errors :) 
I gave up fixing it because it required extra refactoring (and to be brutally honest I also got lazy because I didnt want to deal with it). I can take a stab and fix it though if you feel we should start making that effort to make our agent better.

Reply:
Should we move this function to `future.py`? `textutil` seems to have been designed as a common utility class, so ideally it shouldn't be relying on anything else in the agent.

Reply:
I totally understand your concern but a lot of different functions in textutil have similar checks for py2 :)

Apart from the py-version check I dont have any other dependency and since this is a function that deals with strings in general I feel textutil is better suited for the job. 

But yeah if you guys feel strongly about it being in future.py, I'll move it not a big deal

Reply:
Ahh, I just noticed we use inline imports for `PY_VERSION_MAJOR` all over the place in this module. Fun! :smile:

Reply:
Actually, we mix it up... function `gen_password_hash` in this module is querying `sys.version_info` instead of relying on `PY_VERSION_MAJOR` (which does the same thing). We will eventually need to address this since I'm guessing the linter will complain about imports not being at the top of the file and then either break the circular dependency or remove `PY_VERSION_MAJOR` and use `sys.version_info` directly.

Up to you if you want to take this up or not.

Reply:
sounds good